src/ripple/core/SQLInterface.h

```diff
@@ -0,0 +1,259 @@
+        {
+            auto hash1 = std::hash<T1>{}(p.first);
+            auto hash2 = std::hash<T2>{}(p.second);
+            return hash1 ^ hash2;
+        }
+    };
+
+    static std::string
+    tableName(TableType type)
+    {
+        switch (type)
+        {
+            case LEDGERS:
+                return "Ledgers";
+            case TRANSACTIONS:
+                return "Transactions";
+            case ACCOUNT_TRANSACTIONS:
+                return "AccountTransactions";
+        }
+        return "Unknown";
+    }
+
+    static std::unique_ptr<SQLInterface>
+    init(Application& app, Config const& config);
+
+    virtual bool
+    makeLedgerDBs(
+        Config const& config,
+        beast::Journal const& j,
+        bool setupFromConfig,
+        LedgerIndex shardIndex,
+        bool backendComplete,
+        boost::filesystem::path const& dir,
+        DatabaseCon::CheckpointerSetup const& checkpointerSetup) = 0;
+
+    virtual void
+    addDBs(LedgerIndex shardIndex, DatabaseCon* tx, DatabaseCon* lgr) = 0;
+
+    virtual void
+    removeDBs(LedgerIndex shardIndex) = 0;
+
+    virtual ~SQLInterface()
+    {
+    }
+
+    virtual boost::optional<LedgerIndex>
+    getMinLedgerSeq(DatabaseType dbtype, TableType type) = 0;
+
+    virtual boost::optional<LedgerIndex>
+    getMaxLedgerSeq(DatabaseType dbtype, TableType type) = 0;
+
+    virtual void
+    deleteByLedgerSeq(
+        DatabaseType dbtype,
+        TableType type,
+        LedgerIndex ledgerSeq) = 0;
+
+    virtual void
+    deleteBeforeLedgerSeq(
+        DatabaseType dbtype,
+        TableType type,
+        LedgerIndex ledgerSeq) = 0;
+
+    virtual int
+    getRows(DatabaseType dbtype, TableType type) = 0;
+
+    virtual std::tuple<int, int, int>
+    getRowsMinMax(DatabaseType dbtype, TableType type) = 0;
+
+    virtual bool
+    saveValidatedLedger(
+        Application& app,
+        std::shared_ptr<Ledger const> const& ledger,
+        bool current) = 0;
+
+    virtual bool
+    loadLedgerInfoByIndex(
+        SQLInterface::SQLLedgerInfo& info,
+        beast::Journal const& j,
+        LedgerIndex ledgerSeq) = 0;
+
+    virtual bool
+    loadLedgerInfoByIndexSorted(
+        SQLInterface::SQLLedgerInfo& info,
+        beast::Journal const& j,
+        bool ascendSort) = 0;
+
+    virtual bool
+    loadLedgerInfoByIndexLimitedSorted(
+        SQLInterface::SQLLedgerInfo& info,
+        beast::Journal const& j,
+        LedgerIndex ledgerFirstIndex,
+        bool ascendSort) = 0;
+
+    virtual bool
+    loadLedgerInfoByHash(
+        SQLInterface::SQLLedgerInfo& info,
+        beast::Journal const& j,
+        uint256 const& ledgerHash) = 0;
+
+    virtual uint256
+    getHashByIndex(LedgerIndex ledgerIndex) = 0;
+
+    virtual bool
+    getHashesByIndex(
+        beast::Journal const& j,
+        LedgerIndex ledgerIndex,
+        uint256& ledgerHash,
+        uint256& parentHash) = 0;
+
+    virtual std::map<LedgerIndex, std::pair<uint256, uint256>>
```

Comment:
The pair with identical types is error prone to use:
1) There no way to know that whose two hases are the ledger hash and the parent hash without looking at the implementation
2) Even if we knew, it's very easy to accidentally swat the values. Consider using a struct.

Reply:
Done.